wallet/key_deriver.go

```diff
@@ -0,0 +1,153 @@
+)
+
+type KeyDeriver struct {
+	privateKey *ec.PrivateKey
+	publicKey  *ec.PublicKey
+}
+
+func NewKeyDeriver(privateKey *ec.PrivateKey) *KeyDeriver {
+	return &KeyDeriver{
+		privateKey: privateKey,
+		publicKey:  privateKey.PubKey(),
+	}
+}
+
+func (kd *KeyDeriver) DeriveSymmetricKey(protocol WalletProtocol, keyID string, counterparty WalletCounterparty) ([]byte, error) {
+	// Prevent deriving symmetric key for self
+	if counterparty.Type == CounterpartyTypeSelf {
+		return nil, fmt.Errorf("cannot derive symmetric key for self")
+	}
+
+	// If counterparty is 'anyone', use a fixed public key
+	if counterparty.Type == CounterpartyTypeAnyone {
+		_, fixedKey := ec.PrivateKeyFromBytes([]byte{1})
+		counterparty = WalletCounterparty{
+			Type:         CounterpartyTypeOther,
+			Counterparty: fixedKey,
+		}
+	}
+
+	// Derive both public and private keys
+	derivedPublicKey, err := kd.DerivePublicKey(protocol, keyID, counterparty, false)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive public key: %w", err)
+	}
+
+	derivedPrivateKey, err := kd.DerivePrivateKey(protocol, keyID, counterparty)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive private key: %w", err)
+	}
+
+	// Create shared secret
+	sharedSecret, err := derivedPrivateKey.DeriveSharedSecret(derivedPublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create shared secret: %w", err)
+	}
+	if sharedSecret == nil {
+		return nil, fmt.Errorf("failed to derive shared secret")
+	}
+
+	// Return the x coordinate of the shared secret point
+	return sharedSecret.X.Bytes(), nil
+}
+
+func (kd *KeyDeriver) DerivePublicKey(protocol WalletProtocol, keyID string, counterparty WalletCounterparty, forSelf bool) (*ec.PublicKey, error) {
+	counterpartyKey := kd.NormalizeCounterparty(counterparty)
+	invoiceNumber, err := kd.ComputeInvoiceNumber(protocol, keyID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compute invoice number: %w", err)
+	}
+
+	var pubKey *ec.PublicKey
+	if forSelf {
+		privKey, err := kd.privateKey.DeriveChild(counterpartyKey, invoiceNumber)
+		if err != nil {
+			return nil, fmt.Errorf("failed to derive child private key: %w", err)
+		}
+		pubKey = privKey.PubKey()
+	} else {
+		pubKey, err = counterpartyKey.DeriveChild(kd.privateKey, invoiceNumber)
+		if err != nil {
+			return nil, fmt.Errorf("failed to derive child public key: %w", err)
+		}
+	}
+	return pubKey, nil
+}
+
+func (kd *KeyDeriver) DerivePrivateKey(protocol WalletProtocol, keyID string, counterparty WalletCounterparty) (*ec.PrivateKey, error) {
+	counterpartyKey := kd.NormalizeCounterparty(counterparty)
+	invoiceNumber, err := kd.ComputeInvoiceNumber(protocol, keyID)
+	if err != nil {
+		return nil, err
+	}
+
+	k, err := kd.privateKey.DeriveChild(counterpartyKey, invoiceNumber)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive child key: %w", err)
+	}
+	return k, nil
+}
+
+func (kd *KeyDeriver) NormalizeCounterparty(counterparty WalletCounterparty) *ec.PublicKey {
```

Comment:
Maybe this one doesn't need to be public?

Reply:
You are correct that this is `private` in `ts-sdk`, so would be good to match. This is used in testing so we would have to switch from blackbox test to whitebox, which is fine.